Mathlib/Tactic/Find.lean

```diff
@@ -39,96 +102,249 @@ private partial def matchHyps : List Expr → List Expr → List Expr → MetaM
+  DeclCache.mk
+    (profilingName := "#find: init cache")
+    (empty := ({}, {}))
+    (addLibraryDecl := fun _ c (m₁, m₂) ↦ return (← NameRel.addDecl c m₁, m₂))
+    (addDecl := fun _ c (m₁, m₂) ↦ return (m₁, ← NameRel.addDecl c m₂))
+
+-- NB: In large files it may be slightly wasteful to calculate a full NameSet for the local
+-- definition upon every invocation of `#find`, and a linear scan might work better. For now,
+-- this keeps the code below more uniform.
+
+/-- The parsed and elaborated arguments to `#find`  -/
+structure Arguments where
+  /-- Identifiers to search for -/
+  idents : Array Name
+  /-- Lemma name substrings to search for -/
+  name_pats : Array String
+  /-- Term patterns to search for. The boolean indicates conclusion patterns. -/
+  terms : Array (Bool × Expr)
+
+/-- The core of the `#find` tactic with all parsing/printing factored out, for
+programmatic use.  -/
+def find (args : Arguments) (maxShown := 200) :
+  MetaM (Except MessageData (MessageData × List Name)) := do
+  profileitM Exception "#find" (← getOptions) do
+    -- Collect set of names to query the index by
+    let needles : NameSet :=
+          {} |> args.idents.foldl NameSet.insert
+             |> args.terms.foldl (fun s (_,t) => t.foldConsts s (flip NameSet.insert))
+    if needles.isEmpty then do
+      return .error m!"Cannot search: No constants in search pattern."
+
+    -- Prepare term patterns
+    let pats ← liftM $ args.terms.mapM fun (is_conclusion_pattern, t) =>
+      if is_conclusion_pattern
+      then matchConclusion t
+      else matchAnywhere t
+
+    -- Query the declaration cache
+    let (m₁, m₂) ← findDeclsByConsts.get
+    let hits := NameSet.intersects $ needles.toArray.map $ fun needle =>
+      NameSet.union (m₁.findD needle {}) (m₂.findD needle {})
+
+    -- Filter by name patterns
+    let hits2 := hits.toArray.filter fun n => args.name_pats.all fun p =>
+      p.isInfixOf n.toString
+
+    -- Filter by term patterns
+    let hits3 ← hits2.filterM fun n => do
+      let env ← getEnv
+      if let some ci := env.find? n then do pats.allM (· ci)
+      else return false
+
+    -- Sort to bring lemmas with small names to the top, and keep related lemmas together
+    let hits4 := hits3.qsort Name.lt
+
+    let summary ← IO.mkRef MessageData.nil
+    let add_line line := do summary.set $ (← summary.get) ++ line ++ Format.line
+
+    let needles_list := MessageData.andList (needles.toArray.map .ofName)
+    add_line $ m!"Found {hits.size} definitions mentioning {needles_list}."
+    unless (args.name_pats.isEmpty) do
+      let name_list := MessageData.andList <| args.name_pats.map fun s => m!"\"{s}\""
+      add_line $ m!"Of these, {hits2.size} have a name containing {name_list}."
+    unless (pats.isEmpty) do
+      add_line $ m!"Of these, {hits3.size} match your patterns."
+    unless (hits4.size ≤ maxShown) do
+      add_line $ m!"Of these, the first {maxShown} are shown."
```

Comment:
```suggestion
    let addLine line := do summary.set <| (← summary.get) ++ line ++ Format.line

    let needlesList := MessageData.andList (needles.toArray.map .ofName)
    addLine m!"Found {hits.size} definitions mentioning {needlesList}."
    unless (args.namePats.isEmpty) do
      let nameList := MessageData.andList <| args.namePats.map fun s => m!"\"{s}\""
      addLine $ m!"Of these, {hits2.size} have a name containing {nameList}."
    unless (pats.isEmpty) do
      addLine $ m!"Of these, {hits3.size} match your patterns."
    unless (hits4.size ≤ maxShown) do
      addLine $ m!"Of these, only the first {maxShown} are shown."
```
Only style comments here, but: I wonder if these informative messages should be hidden behind `#find?` syntax. I can imagine that if many lemmas are matched at each stage, the infoview might become very crowded.